Create @frontity/google-analytics package

DAreRodz opened the pull request:
<!--
Thanks for your pull request 😊. Note that not following the template might result in your issue being closed
-->

#### Description of what you did:

<!-- Please describe everything as much as possible -->
This package will allow any frontity project to integrate with Google Analytics.

~Note: work is not completed yet, this PR is intended to be merged with beta.~

#### My PR is a:

<!-- Delete the ones that don't apply -->

- 🚀 New feature

#### Is the PR ready to be merged?

<!-- Delete the one that don't apply -->

- ✅ Yes!
<!-- - 🚧 Work in progress -->

DAreRodz committed:
Create first analytics packages

DAreRodz committed:
Change typing system

DAreRodz committed:
Implement actions and root

DAreRodz committed:
Implement afterCSR action

DAreRodz committed:
Implement TitleObserver and use it in afterCSR

DAreRodz committed:
Temporarily change state.analytics.namespaces type

DAreRodz committed:
Delete TitleObserver and use  setTimeout

luisherranz commented on packages/analitics/package.json:
```diff
@@ -0,0 +1,34 @@
+{
+  "name": "@frontity/analytics",
+  "version": "0.1.0",
+  "description": "Analytics utilities for »Frontity",
```
I'd say more something like: `This package contains the base types and actions to build analytics packages for Frontity`

DAreRodz replied:
Much better!

luisherranz commented on packages/analitics/types.ts:
```diff
@@ -0,0 +1,40 @@
+  category: string;
+  action: string;
+  label: string;
+};
```
I'm not sure if we should use `action`, `category` and  `label`. Feels too coupled to Google Analytics. Most analytics tools have something like `name` and `payload` for events.

luisherranz commented on packages/analitics/types.ts:
```diff
@@ -0,0 +1,40 @@
+      namespaces: Record<string, string>;
+    };
+    router?: {
+      link: string;
```
I'm yet not sure if it's a better idea to hardcode external types like this (for the sake of simplicity) or to recommend importing the original type like this.

```js
router?: {
  link: Router["state"]["router"]["link"];
}
```

One scenario when this could be relevant is if you are updating this package to be used with `@frontity/router@^2.0.0` and in that version `router` has changed `link` from a string to an object for example. It's not a big deal.

Any other scenarios where this could be more important?

luisherranz commented on packages/analitics/index.ts:
```diff
@@ -0,0 +1,49 @@
+import { observe } from "frontity";
+import Analytics from "./types";
+
```
I think that apart from the named exports, I would also export the default export (containing `state` and `actions`) in this file.

Did you think about a reason not to do so? Do you think it can be misleading?

DAreRodz replied:
I wanted to clearly show that this is not a package but a library, just looking at the code.

luisherranz commented on packages/analitics/index.ts:
```diff
@@ -0,0 +1,49 @@
+       * Execute action in the next tick so we can
+       * ensure the title has changed by React.
+       */
+      setTimeout(() => {
```
I think this can be kind of dangerous... How can we make sure this is not going to break? What happens if React takes more time to update the `document.title`? Maybe we need to think of a better technique to extract the title.

DAreRodz replied:
Here I (hastily) assumed that React would render syncronously as soon as `data.isReady` was `true`, so it would be `safe` to use `setTimeout` in this case.

Maybe I assumed wrong 😅

luisherranz commented on packages/google-analytics/src/index.ts:
```diff
@@ -0,0 +1,62 @@
+  },
+  state: {
+    analytics: {
+      namespaces: { ga: "ga" }
```
What's the advantage of using an object over an array `["ga"]`. Is it just to avoid [this bug](https://github.com/frontity/frontity/issues/237) or does it have a real advantage?

DAreRodz replied:
It's just to avoid the bug, ofc.

luisherranz commented on packages/google-analytics/src/index.ts:
```diff
@@ -0,0 +1,62 @@
+      sendEvent,
+      afterCSR
+    },
+    ga: {
```
I'm not sure if the namespaces should be `googleAnalytics`, `gtmAnalytics`, `mixpanelAnalytics`, `amplitudeAnalytics` and so on... thoughts?

DAreRodz replied:
Using `ga` as the namespace didn't convince me. I prefer your idea. 👍

luisherranz commented on packages/google-analytics/src/components/index.tsx:
```diff
@@ -0,0 +1,28 @@
+
+  // Render Tracker code.
+  return (
+    <script
```
It seems that this script should go in the `<head>` section:
> Copy and paste the global site tag to each web page you want to measure. Add the tag right after the opening <head> tag on each page.

https://support.google.com/analytics/answer/1008080?hl=en

DAreRodz replied:
Good catch 👍

luisherranz commented:
I've made a bunch of comments but it looks great David. I see you've tried to use `MutationObserver` to extract the `document.title`. Maybe we can have a quick call to discuss that.

luisherranz commented:
**Important:** I've changed the base of this PR to `beta`. Please rebase the branch.

DAreRodz committed:
Change frontity/analytics description

DAreRodz committed:
Replace afterCSR by a Root component

Also:
- Use Head in the Google Analytics' Root component
- Rename "ga" namespace to "googleAnalytics"

DAreRodz committed:
Fix typo in the analytics folder name

DAreRodz committed:
Force title changes when isReady becomes true

DAreRodz committed:
Add simple tests to the analytics library

DAreRodz committed:
Add empty tests to google analytics package

DAreRodz committed:
Start with e2e analytics package

DAreRodz committed:
Prevent pageviews to be sent twice

DAreRodz committed:
Update package-lock

DAreRodz committed:
Added changesets for analytics packages

DAreRodz committed:
Fix source version in analytics

DAreRodz committed:
Remove some console logs

DAreRodz committed:
Fix the ubuntu version used in the cypress github action (#279)

* E2E: Add debug env variable

* Change ubuntu version to 16.04

* Remove DEBUG env variable

Co-authored-by: Luis Herranz <luisherranz@gmail.com>

DAreRodz committed:
Change package-lock again

DAreRodz committed:
Merge branch 'beta' into analytics
